### PR TITLE
WIP: Commit API and code refactor + Unit Tests

### PR DESCRIPTION
**Pull request by aakashns**
This PR aims to simply the `jovian.commit` API by renaming, removing and adding some arguments, as well as handling some corner cases to improve the user experience & address common exceptions. This will also prepare the library for integration with `papermill`. All changes will be backward compatible.

Work items:
- [x] Extract helper functions from `jovian.commit`
- [x] Move `commit` & helper functions to a separate file
- [x] Move record logging functions to separate file
- [x] Rename `jovian.commit` arguments for clarity
- [x] Support `jovian clone` without outputs
- [x] Provide arguments to set gist title & version title directly from `jovian.commit`
- [x] Ensure backwards compatibility using `**kwargs` (old arguments can be still be used with deprecation warning)
- [x] Update library callbacks to use new API
- [x] Manual testing of all new/modified functionality
- [ ] Write unit tests for all files
- [ ] Set up automatic test evaluation before publish

Testing checklist:
- [ ] `jovian/_flavor.py`
- [ ] `jovian/_version.py`
- [ ] `jovian/callbacks/fastai.py`
- [ ] `jovian/callbacks/keras.py`
- [ ] `jovian/callbacks/__init__.py`
- [ ] `jovian/__init__.py`
- [ ] `jovian/utils/misc.py`
- [ ] `jovian/utils/git.py`
- [ ] `jovian/utils/rcfile.py`
- [ ] `jovian/utils/configure.py`
- [ ] `jovian/utils/error.py`
- [ ] `jovian/utils/initialize.py`
- [ ] `jovian/utils/credentials.py`
- [ ] `jovian/utils/records.py`
- [ ] `jovian/utils/constants.py`
- [ ] `jovian/utils/request.py`
- [ ] `jovian/utils/commit.py`
- [ ] `jovian/utils/__init__.py`
- [ ] `jovian/utils/logger.py`
- [ ] `jovian/utils/latest.py`
- [ ] `jovian/utils/api.py`
- [ ] `jovian/utils/extension.py`
- [ ] `jovian/utils/environment.py`
- [ ] `jovian/utils/install.py`
- [ ] `jovian/utils/script.py`
- [ ] `jovian/utils/slack.py`
- [ ] `jovian/utils/clone.py`
- [ ] `jovian/utils/jupyter.py`
- [ ] `jovian/utils/envfile.py`
- [ ] `jovian/cli.py`

[ch1483]

### CODE REVIEW

File: jovian/tests/utils/test_latest.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+from unittest import TestCase, mock
+from jovian._version import __version__
+from pkg_resources import parse_version
+
+from jovian.utils.latest import _get_latest_version, _print_update_message, check_update
+
+
+class TestGetLatestVersion(TestCase):
+    def error_raiser(*args, **kwargs):
+        raise ImportError('fake requests.get error')
```
Comment (aakashns) [inline jovian/tests/utils/test_latest.py]: need to change the `Error` type

File: jovian/tests/utils/test_latest.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from unittest import TestCase, mock
+from jovian._version import __version__
+from pkg_resources import parse_version
+
+from jovian.utils.latest import _get_latest_version, _print_update_message, check_update
+
+
+class TestGetLatestVersion(TestCase):
```
Comment (aakashns) [inline jovian/tests/utils/test_latest.py]: also test the case inside `try` by mocking `requests.get`

File: jovian/tests/utils/test_latest.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    @mock.patch("jovian.utils.latest.requests.get", mock.Mock(return_value=error_raiser))
+    def test_get_latest_version_exception(self):
+        expected_result = __version__
+        self.assertEqual(_get_latest_version(), expected_result)
+
+
+class TestPrintUpdateMessage(TestCase):
+
+    @mock.patch("jovian.utils.latest.log")
+    def test_print_update_message_normal(self, mock_log):
```
Comment (aakashns) [inline jovian/tests/utils/test_latest.py]: can directly test the output here from stdout. Check `test_logger.py`

File: jovian/tests/utils/test_latest.py (hunk 1 above)
Comment (rohitsanj) [reply jovian/tests/utils/test_latest.py]: changed to `ConnectionError`

File: jovian/tests/utils/test_misc.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        self.assertEqual(get_flavor(), "jovian")
+
+    def test_get_flavor_import_error(self):
+        import builtins
+        _original_import = builtins.__import__
+
+        def error_raiser(*args, **kwargs):
+            raise ImportError('fake import error')
+
+        builtins.__import__ = error_raiser
```
Comment (rohitsanj) [inline jovian/tests/utils/test_misc.py]: We can put the import in a `mock.patch` decorator like 
```python
@mock.patch("builtins.__import__", mock.Mock(return_value=ImportError('fake import error'))
```

I've done that in some tests.. what do you think?

**Comment by aakashns** [comment]
I have tested the top-level API manually, and everything is working as expected. We will publish this version, and continue the unit testing in a new branch.
